File: src/alignment_filter_node.py
#!/usr/bin/env python3

import rospy
from nav_msgs.msg import Odometry, Path
from geometry_msgs.msg import PoseStamped, TransformStamped
from std_msgs.msg import Float32MultiArray
from evo.core.trajectory import PoseTrajectory3D
from ais_robot_localization.msg import LocalizationMonitorResult
from utils import odom_to_se3, se3_to_posestamped, snake_alignment, gaussian_weight, euclidean_distance
import threading
import numpy as np
import tf.transformations as tf
import tf2_ros
from geometry_msgs.msg import Quaternion

class AlignmentBasedFilterNode:
    def __init__(self):
        # Initialize the node
        rospy.init_node('alignment_filter_node')

        self.publish_tf = rospy.get_param('~publish_tf', True)

        # Initialize data storage
        self.data_list = []
        self.error_trans_avg = []
        self.error_rot_avg = []
        self.distances_current_estimate = []
        self.time_stamp_list = []
        self.global_path = Path()
        self.global_poses = []
        self.local_poses = []
        self.local_path = Path()
        self.local_path_transformed = Path()
        self.used_length = 0.0

        self.current_odom_state = None
        
        # Frame IDs for global and local paths
        self.global_frame = "map"
        self.local_frame = "odom"
        self.last_selected = -float('inf')

        self.current_transform = None

        # Transform broadcaster
        if self.publish_tf:
            self.tf_broadcaster = tf2_ros.TransformBroadcaster()
        
        # Publishers for path visualization
        self.global_path_pub = rospy.Publisher('/alignment_global_path', Path, queue_size=10)
        self.local_path_transformed_pub = rospy.Publisher('/alignment_local_path_transformed', Path, queue_size=10)
        self.odom_publisher = rospy.Publisher('/alignment_odometry', Odometry, queue_size=10)
        
        # Subscriber for LocalizationMonitorResult messages
        self.sub = rospy.Subscriber('/localization_result', LocalizationMonitorResult, self.localization_monitor_callback)

        # Initialize lock for thread safety
        self.lock = threading.Lock()

        self.local_odom_sub = rospy.Subscriber("/local_odom", Odometry, self.local_odom_callback)

        rospy.loginfo("AlignmentBasedFilterNode initialized")


    def publish_transform(self, odom):
        """Publish the inverse of the current transformation as map -> odom."""
        if self.current_transform is None:
            return

        try:
            # Compute the inverse transform
            transform = self.current_transform#tf.inverse_matrix(self.current_transform)

            # Create a TransformStamped message
            t = TransformStamped()
            t.header.stamp = odom.header.stamp
            t.header.frame_id = self.global_frame  # Parent frame
            t.child_frame_id = self.local_frame    # Child frame


            # Extract translation and rotation
            translation = transform[:3, 3]
            rotation_matrix = transform[:3, :3]

            # Convert rotation matrix to quaternion
            quaternion = tf.quaternion_from_matrix(transform)

            # Fill the TransformStamped message
            t.transform.translation.x = translation[0]
            t.transform.translation.y = translation[1]
            t.transform.translation.z = translation[2]
            t.transform.rotation.x = quaternion[0]
            t.transform.rotation.y = quaternion[1]
            t.transform.rotation.z = quaternion[2]
            t.transform.rotation.w = quaternion[3]

            # Broadcast the transform
            self.tf_broadcaster.sendTransform(t)
            #rospy.loginfo("Published transform: map -> odom")
        except Exception as e:
            rospy.logerr(f"Error publishing transform: {e}")


    def local_odom_callback(self, odom_msg):
        with self.lock:
            # Extract the current position and orientation from the Odometry message
            position = np.array([odom_msg.pose.pose.position.x,
                                odom_msg.pose.pose.position.y,
                                odom_msg.pose.pose.position.z, 1])
            
            # Convert quaternion to a 3x3 rotation matrix
            orientation_q = odom_msg.pose.pose.orientation
            rotation_matrix = tf.quaternion_matrix([orientation_q.x, orientation_q.y, orientation_q.z, orientation_q.w])
            
            # Create the current pose 4x4 matrix
            current_pose_matrix = np.eye(4)
            current_pose_matrix[:3, :3] = rotation_matrix[:3, :3]
            current_pose_matrix[:3, 3] = position[:3]
            
            # Apply the transformation
            if self.current_transform is None or self.used_length < 10.0:
                transformed_pose_matrix = current_pose_matrix
            else:
                if self.publish_tf:
                    self.publish_transform(odom_msg)
                transformed_pose_matrix = np.dot(self.current_transform, current_pose_matrix)
            
            # Extract the transformed position
            transformed_position = transformed_pose_matrix[:3, 3]
            
            # Convert the transformed rotation matrix back to a quaternion
            transformed_quaternion = tf.quaternion_from_matrix(transformed_pose_matrix)
            
            # Update the Odometry message
            transformed_odom = Odometry()
            transformed_odom.header = odom_msg.header  # Keep the same header
            transformed_odom.child_frame_id = odom_msg.child_frame_id
            transformed_odom.header.frame_id = self.global_frame
            
            # Update position
            transformed_odom.pose.pose.position.x = transformed_position[0]
            transformed_odom.pose.pose.position.y = transformed_position[1]
            transformed_odom.pose.pose.position.z = transformed_position[2]
            
            # Update orientation
            transformed_odom.pose.pose.orientation = Quaternion(
                x=transformed_quaternion[0],
                y=transformed_quaternion[1],
                z=transformed_quaternion[2],
                w=transformed_quaternion[3]
            )

            self.current_odom_state = transformed_odom
            self.odom_publisher.publish(transformed_odom)


    def localization_monitor_callback(self, msg):
        with self.lock:

            # Extract frames from the first message and set them
            if not self.global_frame:
                self.global_frame = msg.pose_global.header.frame_id
            if not self.local_frame:
                self.local_frame = msg.pose_local.header.frame_id

            features = msg.float_array

            # Store the message in the data list
            self.data_list.append(msg)

            # Add global pose to the global path
            global_pose_stamped = PoseStamped()
            global_pose_stamped.header = msg.pose_global.header
            global_pose_stamped.pose = msg.pose_global.pose
            self.global_path.poses.append(global_pose_stamped)
            self.global_poses.append(odom_to_se3(msg.pose_global.pose))
            self.time_stamp_list.append(msg.pose_global.header.stamp)

            # Add local pose to the local path
            local_pose_stamped = PoseStamped()
            local_pose_stamped.header = msg.pose_local.header
            local_pose_stamped.pose = msg.pose_local.pose
            self.local_path.poses.append(local_pose_stamped)
            self.local_poses.append(odom_to_se3(msg.pose_local.pose))

            self.error_trans_avg.append(features[0])

            #if self.current_transform is None:
            self.distances_current_estimate.append(0.0)
            #else:
                #self.distances_current_estimate.append(euclidean_distance(msg.pose_global.pose, self.current_odom_state.pose.pose))

            rospy.loginfo("Added new global and local poses to paths with error %f^, distance to current estimate %f", self.error_trans_avg[-1], self.distances_current_estimate[-1])
            
            self.clean_up()
            if(len(self.data_list) > 3):
                self.align_local_to_global_odom_and_calc_errors(msg.pose_global.header)
                self.publish_paths()
                self.used_length = self.data_list[-1].cumulative_length - self.data_list[0].cumulative_length

    def clean_up(self):
        length = self.data_list[-1].cumulative_length - self.data_list[0].cumulative_length
        rospy.loginfo(length)
        while (self.data_list[-1].cumulative_length - self.data_list[0].cumulative_length) > 150:
            self.data_list.pop(0)
            self.time_stamp_list.pop(0)
            self.global_path.poses.pop(0)
            self.global_poses.pop(0)
            self.local_poses.pop(0)
            self.local_path.poses.pop(0)
            self.error_trans_avg.pop(0)
            self.distances_current_estimate.pop(0)



    def align_local_to_global_odom_and_calc_errors(self, header):
        rospy.loginfo("Align local´odom path to global odom path")
        self.local_odom_path = Path()

        local_trajectory = PoseTrajectory3D(poses_se3=np.copy(self.local_poses), timestamps=self.time_stamp_list)
        local_trajectory_no_snake = PoseTrajectory3D(poses_se3=np.copy(self.local_poses), timestamps=self.time_stamp_list)
        global_trajectory = PoseTrajectory3D(poses_se3=np.copy(self.global_poses), timestamps=self.time_stamp_list)

        #error_mean = np.square(np.mean(self.error_trans_avg))
        #error_distance_squared = np.square(np.array(self.distances_current_estimate))
        #error_distance_squared_mean = np.mean(error_distance_squared)

        penalty_values = np.array(self.error_trans_avg)
        #if(self.used_length > 30.0):
        #    penalty_values += error_distance_squared

        ref_value =  np.percentile(penalty_values, 25)

        weights = gaussian_weight(penalty_values, ref_value)
        rospy.loginfo("ref_val: %f, min weight: %f, max weigth: %f", ref_value, np.min(weights), np.max(weights))
        self.current_transform = snake_alignment(local_trajectory, global_trajectory, weights=weights)

        local_trajectory_no_snake.transform(self.current_transform)
        self.local_path_transformed.poses.clear()      
        for i, pose_se3 in enumerate(local_trajectory_no_snake.poses_se3):
            pose_stamped = se3_to_posestamped(pose_se3, self.time_stamp_list[i], z_to_zero=False)
            pose_stamped.header = header
            pose_stamped.header.frame_id = self.global_frame
            self.local_path_transformed.poses.append(pose_stamped)

    def publish_paths(self):
        # Update path headers with the correct frames and publish
        self.global_path.header.frame_id = self.global_frame
        self.global_path.header.stamp = rospy.Time.now()
        self.global_path_pub.publish(self.global_path)
        
        self.local_path_transformed.header.frame_id = self.global_frame
        self.local_path_transformed.header.stamp = rospy.Time.now()
        self.local_path_transformed_pub.publish(self.local_path_transformed)

        rospy.loginfo("Published global and local paths with %d and %d poses", len(self.global_path.poses), len(self.local_path.poses))

if __name__ == '__main__':
    try:
        node = AlignmentBasedFilterNode()
        rospy.spin()
    except rospy.ROSInterruptException:
        rospy.loginfo("AlignmentBasedFilterNode has been interrupted.")
    except Exception as e:
        rospy.logerr(f"Unexpected error: {e}")

File: src/alignment_filter_node.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <functional>
#include <memory>
#include <mutex>
#include <string>
#include <utility>
#include <vector>

#include <Eigen/Geometry>

#include <geometry_msgs/msg/pose.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <nav_msgs/msg/path.hpp>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/header.hpp>
#include <tf2/exceptions.h>
#include <tf2/time.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/transform_listener.h>

#include "ais_robot_localization/alignment_filter.hpp"
#include "ais_robot_localization/se3_algorithms.hpp"
#include "robot_localization/msg/localization_monitor_result.hpp"

namespace ais_robot_localization
{
namespace
{

Eigen::Quaterniond toEigenQuaternion(const geometry_msgs::msg::Quaternion & q_msg)
{
  Eigen::Quaterniond q(q_msg.w, q_msg.x, q_msg.y, q_msg.z);
  if (q.norm() == 0.0) {
    return Eigen::Quaterniond::Identity();
  }
  q.normalize();
  return q;
}

Eigen::Isometry3d projectToXYPlane(const Eigen::Isometry3d & transform)
{
  Eigen::Isometry3d projected = Eigen::Isometry3d::Identity();
  projected.translation().x() = transform.translation().x();
  projected.translation().y() = transform.translation().y();
  projected.translation().z() = 0.0;

  const Eigen::Matrix3d & rotation = transform.linear();
  double yaw = std::atan2(rotation(1, 0), rotation(0, 0));
  projected.linear() = Eigen::AngleAxisd(yaw, Eigen::Vector3d::UnitZ()).toRotationMatrix();
  return projected;
}

void projectTrajectoryToXYPlane(std::vector<Eigen::Isometry3d> & trajectory)
{
  for (auto & pose : trajectory) {
    pose = projectToXYPlane(pose);
  }
}

Eigen::Isometry3d poseMsgToIsometry(const geometry_msgs::msg::Pose & pose_msg)
{
  Eigen::Vector3d position(pose_msg.position.x, pose_msg.position.y, pose_msg.position.z);
  Eigen::Quaterniond orientation = toEigenQuaternion(pose_msg.orientation);
  return odomToSE3(position, orientation);
}

geometry_msgs::msg::PoseStamped toRosPoseStamped(
  const PoseStamped & pose,
  const std::string & frame_id,
  const rclcpp::Time & stamp)
{
  geometry_msgs::msg::PoseStamped msg;
  msg.header.frame_id = frame_id;
  msg.header.stamp = stamp;
  msg.pose.position.x = pose.position.x();
  msg.pose.position.y = pose.position.y();
  msg.pose.position.z = pose.position.z();
  msg.pose.orientation.x = pose.orientation.x();
  msg.pose.orientation.y = pose.orientation.y();
  msg.pose.orientation.z = pose.orientation.z();
  msg.pose.orientation.w = pose.orientation.w();
  return msg;
}

geometry_msgs::msg::Pose toRosPose(const Eigen::Isometry3d & transform)
{
  geometry_msgs::msg::Pose pose_msg;
  Eigen::Quaterniond q(transform.linear());
  q.normalize();
  pose_msg.position.x = transform.translation().x();
  pose_msg.position.y = transform.translation().y();
  pose_msg.position.z = transform.translation().z();
  pose_msg.orientation.x = q.x();
  pose_msg.orientation.y = q.y();
  pose_msg.orientation.z = q.z();
  pose_msg.orientation.w = q.w();
  return pose_msg;
}

rclcpp::Time secondsToTime(double stamp_sec)
{
  return rclcpp::Time(static_cast<int64_t>(stamp_sec * 1e9));
}

}  // namespace

class AlignmentFilterNode : public rclcpp::Node
{
public:
  AlignmentFilterNode()
  : rclcpp::Node("alignment_filter_node"),
    publish_tf_(this->declare_parameter("publish_tf", true)),
    two_d_mode_(this->declare_parameter("2D_mode", false)),
    current_transform_(Eigen::Isometry3d::Identity())
  {
    global_frame_ = this->declare_parameter<std::string>("global_frame", "map");
    local_frame_ = this->declare_parameter<std::string>("local_frame", "odom");

    global_path_pub_ = this->create_publisher<nav_msgs::msg::Path>("alignment_global_path", 10);
    local_path_transformed_pub_ =
      this->create_publisher<nav_msgs::msg::Path>("alignment_local_path_transformed", 10);
    odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>("alignment_odometry", 10);

    localization_result_sub_ = this->create_subscription<robot_localization::msg::LocalizationMonitorResult>(
      "localization_result", 50,
      std::bind(&AlignmentFilterNode::localizationMonitorCallback, this, std::placeholders::_1));
    local_odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "local_odom", rclcpp::SensorDataQoS(),
      std::bind(&AlignmentFilterNode::localOdomCallback, this, std::placeholders::_1));

    if (publish_tf_) {
      tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
      tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
      tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
      tf_timer_ = this->create_wall_timer(
        std::chrono::milliseconds(200),
        std::bind(&AlignmentFilterNode::publishTransformTimer, this));
    }

    RCLCPP_INFO(this->get_logger(), "AlignmentFilterNode initialized");
  }

private:
  void localizationMonitorCallback(
    const robot_localization::msg::LocalizationMonitorResult::SharedPtr msg)
  {
    std::lock_guard<std::mutex> lock(mutex_);

    if (!msg->pose_global.header.frame_id.empty()) {
      global_frame_ = msg->pose_global.header.frame_id;
    }
    if (!msg->pose_local.header.frame_id.empty()) {
      local_frame_ = msg->pose_local.header.frame_id;
    }

    global_path_.poses.push_back(msg->pose_global);
    local_path_.poses.push_back(msg->pose_local);

    Eigen::Isometry3d global_pose = poseMsgToIsometry(msg->pose_global.pose);
    Eigen::Isometry3d local_pose = poseMsgToIsometry(msg->pose_local.pose);
    double timestamp_sec = rclcpp::Time(msg->pose_global.header.stamp).seconds();
    double cumulative_length = msg->cumulative_length;
    double trans_error = msg->float_array.empty() ? 0.0 : static_cast<double>(msg->float_array.front());

    std::size_t removed = filter_.addMeasurement(global_pose, local_pose, timestamp_sec, trans_error,
        cumulative_length);

    if (removed > 0) {
      if (global_path_.poses.size() > removed) {
        global_path_.poses.erase(global_path_.poses.begin(), global_path_.poses.begin() + removed);
      } else {
        global_path_.poses.clear();
      }

      if (local_path_.poses.size() > removed) {
        local_path_.poses.erase(local_path_.poses.begin(), local_path_.poses.begin() + removed);
      } else {
        local_path_.poses.clear();
      }
    }

    if (filter_.hasSufficientData()) {
      AlignmentResult result;
      if (filter_.computeAlignment(result)) {
        Eigen::Isometry3d transform_to_use = two_d_mode_ ? projectToXYPlane(result.transform) : result.transform;
        if (two_d_mode_) {
          projectTrajectoryToXYPlane(result.transformed_local);
        }
        current_transform_ = transform_to_use;
        updateTransformedPath(result.transformed_local, filter_.timestamps(), msg->pose_global.header);
        publishPaths();
      }
    }
  }

  void localOdomCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    std::lock_guard<std::mutex> lock(mutex_);

    Eigen::Vector3d position(msg->pose.pose.position.x, msg->pose.pose.position.y, msg->pose.pose.position.z);
    Eigen::Quaterniond orientation = toEigenQuaternion(msg->pose.pose.orientation);
    Eigen::Isometry3d current_pose = odomToSE3(position, orientation);
    Eigen::Isometry3d transformed_pose = current_transform_ * current_pose;
    if (two_d_mode_) {
      transformed_pose = projectToXYPlane(transformed_pose);
    }

    auto transformed_msg = *msg;
    transformed_msg.header.frame_id = global_frame_;
    transformed_msg.pose.pose = toRosPose(transformed_pose);
    transformed_msg.pose.covariance = {
      1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
      0.0, 1.0, 0.0, 0.0, 0.0, 0.0,
      0.0, 0.0, 1.0, 0.0, 0.0, 0.0,
      0.0, 0.0, 0.0, 0.1, 0.0, 0.0,
      0.0, 0.0, 0.0, 0.0, 0.1, 0.0,
      0.0, 0.0, 0.0, 0.0, 0.0, 0.1
    };

    odom_pub_->publish(transformed_msg);
  }

  void updateTransformedPath(
    const std::vector<Eigen::Isometry3d> & transformed_local,
    const std::vector<double> & time_stamps,
    const std_msgs::msg::Header & header)
  {
    local_path_transformed_.poses.clear();
    local_path_transformed_.header.frame_id = global_frame_;
    local_path_transformed_.header.stamp = this->now();

    for (std::size_t i = 0; i < transformed_local.size(); ++i) {
      double stamp = (i < time_stamps.size()) ? time_stamps[i] : rclcpp::Time(header.stamp).seconds();
      PoseStamped pose = se3ToPoseStamped(transformed_local[i], stamp, two_d_mode_);
      local_path_transformed_.poses.push_back(
        toRosPoseStamped(pose, global_frame_, secondsToTime(stamp)));
    }
  }

  void publishPaths()
  {
    rclcpp::Time now = this->now();
    global_path_.header.frame_id = global_frame_;
    global_path_.header.stamp = now;
    local_path_transformed_.header.stamp = now;

    global_path_pub_->publish(global_path_);
    local_path_transformed_pub_->publish(local_path_transformed_);
  }

  void publishTransformTimer()
  {
    if (!publish_tf_ || !tf_buffer_ || !tf_listener_ || !tf_broadcaster_) {
      return;
    }

    try {
      geometry_msgs::msg::TransformStamped latest_tf = tf_buffer_->lookupTransform(
        local_frame_, "base_link", tf2::TimePointZero, tf2::durationFromSec(1.0));
      geometry_msgs::msg::TransformStamped transform_msg;
      transform_msg.header.stamp = latest_tf.header.stamp;
      transform_msg.header.frame_id = global_frame_;
      transform_msg.child_frame_id = local_frame_;

      Eigen::Quaterniond q(current_transform_.linear());
      q.normalize();
      Eigen::Vector3d t = current_transform_.translation();

      transform_msg.transform.translation.x = t.x();
      transform_msg.transform.translation.y = t.y();
      transform_msg.transform.translation.z = t.z();
      transform_msg.transform.rotation.x = q.x();
      transform_msg.transform.rotation.y = q.y();
      transform_msg.transform.rotation.z = q.z();
      transform_msg.transform.rotation.w = q.w();

      tf_broadcaster_->sendTransform(transform_msg);
    } catch (const tf2::TransformException & ex) {
      RCLCPP_ERROR_THROTTLE(this->get_logger(), *this->get_clock(), 1000, "Error publishing transform: %s", ex.what());
    }
  }

  bool publish_tf_;
  bool two_d_mode_;

  rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr global_path_pub_;
  rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr local_path_transformed_pub_;
  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;

  rclcpp::Subscription<robot_localization::msg::LocalizationMonitorResult>::SharedPtr
    localization_result_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr local_odom_sub_;

  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
  std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  rclcpp::TimerBase::SharedPtr tf_timer_;

  nav_msgs::msg::Path global_path_;
  nav_msgs::msg::Path local_path_;
  nav_msgs::msg::Path local_path_transformed_;

  std::string global_frame_;
  std::string local_frame_;

  AlignmentFilter filter_;
  Eigen::Isometry3d current_transform_;

  std::mutex mutex_;
};

}  // namespace ais_robot_localization

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ais_robot_localization::AlignmentFilterNode>());
  rclcpp::shutdown();
  return 0;
}

